pkg/syncutil/cachemanager/cachemanager.go

```diff
@@ -3,34 +3,179 @@ package cachemanager
+	if config.WatchedSet == nil {
+		return nil, fmt.Errorf("watchedSet must be non-nil")
+	}
+	if config.Registrar == nil {
+		return nil, fmt.Errorf("registrar must be non-nil")
+	}
+	if config.ProcessExcluder == nil {
+		return nil, fmt.Errorf("processExcluder must be non-nil")
+	}
+	if config.Tracker == nil {
+		return nil, fmt.Errorf("tracker must be non-nil")
+	}
+	if config.Reader == nil {
+		return nil, fmt.Errorf("reader must be non-nil")
+	}
+
 	return &CacheManager{
-		opa:              opa,
-		syncMetricsCache: syncMetricsCache,
-		tracker:          tracker,
-		processExcluder:  processExcluder,
+		opa:                        config.Opa,
+		syncMetricsCache:           config.SyncMetricsCache,
+		tracker:                    config.Tracker,
+		processExcluder:            config.ProcessExcluder,
+		registrar:                  config.Registrar,
+		watchedSet:                 config.WatchedSet,
+		reader:                     config.Reader,
+		specifiedGVKs:              aggregator.NewGVKAggregator(),
+		gvksToList:                 watch.NewSet(),
+		backgroundManagementTicker: *time.NewTicker(3 * time.Second),
+		gvksToDeleteFromCache:      watch.NewSet(),
+	}, nil
+}
+
+func (c *CacheManager) Start(ctx context.Context) error {
+	go c.manageCache(ctx)
+
+	<-ctx.Done()
+	return nil
+}
+
+// AddSource adjusts the watched set of gvks according to the newGVKs passed in
+// for a given sourceKey.
+func (c *CacheManager) AddSource(ctx context.Context, sourceKey aggregator.Key, newGVKs []schema.GroupVersionKind) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if err := c.specifiedGVKs.Upsert(sourceKey, newGVKs); err != nil {
+		return fmt.Errorf("internal error adding source: %w", err)
 	}
+	// as a result of upserting the new gvks for the source key, some gvks
+	// may become unreferenced and need to be deleted; this will be handled async
+	// in the manageCache loop.
+
+	// make changes to the watches
+	if err := c.replaceWatchSet(ctx); err != nil {
+		return fmt.Errorf("error watching new gvks: %w", err)
+	}
+
+	return nil
+}
+
+// replaceWatchSet looks at the specifiedGVKs and makes changes to the registrar's watch set.
+// assumes caller has lock.
+func (c *CacheManager) replaceWatchSet(ctx context.Context) error {
+	newWatchSet := watch.NewSet()
+	newWatchSet.Add(c.specifiedGVKs.ListAllGVKs()...)
+
+	if newWatchSet.Equals(c.watchedSet) {
+		// nothing to do as the sets are equal
+		return nil
+	}
+
+	// record any gvks that need to be deleted
+	c.gvksToDeleteFromCache.AddSet(c.watchedSet.Difference(newWatchSet))
+
+	var innerError error
+	c.watchedSet.Replace(newWatchSet, func() {
+		// *Note the following steps are not transactional with respect to admission control
+
+		// Important: dynamic watches update must happen *after* updating our watchSet.
+		// Otherwise, the sync controller will drop events for the newly watched kinds.
+		// Defer error handling so object re-sync happens even if the watch is hard
```

Comment:
I think so. We are not making a decision on the error at the call site IIUC the comment.

Old code for ref

https://github.com/open-policy-agent/gatekeeper/blob/e60fd23198eda6f746c9263a0fed46423b3a6625/pkg/controller/config/config_controller.go#L296-L310

The only thing we are doing differently is handling the process excluder change separately. 

We are then inspecting the error out of that closure and bubbling it up via `AddSource/ RemoveSource` in our code

Reply:
I can mention retry on failure in the docstring.

I will re-work this comment. I always read it to mean, handle the error, `innerError` outside of the closure `func() { ... } where it's called so that we can act on it. Since if you return the error inside the function it's not possible to return it to the caller. (Also the func signature doesn't allow it).

Reply:
> I will re-work this comment.

IMO the comment can be removed. The purpose of innerError is obvious via its proximity to `return innerError`

Reply:
removed! also added the retry-ness of the error on the docstrings.